block-storage/src/main/scala/coop/rchain/blockstorage/dag/BlockDagFileStorage.scala

```diff
@@ -298,11 +307,23 @@ final class BlockDagFileStorage[F[_]: Concurrent: Sync: Log: RaiseIOError] priva
                                                             BlockSenderIsMalformed(block)
                                                           )
                                                         }
-                deployHashes = deployData(block).map(_.sig).toList
-                _            <- putBlockNumber(block.blockHash, blockNumber(block))
-                _            <- deployIndex.addAll(deployHashes.map(_ -> block.blockHash))
-                _            <- latestMessagesIndex.addAll(newValidatorsWithSenderLatestMessages.toList)
-                _            <- blockMetadataIndex.add(blockMetadata)
+                deployHashes             = deployData(block).map(_.sig).toList
+                sendersSinceLastProposal <- sendersSinceLastProposalIndex.data
+                _                        <- putBlockNumber(block.blockHash, blockNumber(block))
+                _                        <- deployIndex.addAll(deployHashes.map(_ -> block.blockHash))
+                _                        <- latestMessagesIndex.addAll(newValidatorsWithSenderLatestMessages.toList)
+                _                        <- blockMetadataIndex.add(blockMetadata)
+                _                        <- sendersSinceLastProposalIndex.add(block.sender, (block.blockHash, Set.empty))
+                _ <- sendersSinceLastProposal.toList.traverse_ {
+                      case (validator, (_, senders))
+                          if validator == block.sender || senders.contains(block.sender) =>
+                        ().pure[F]
+                      case (validator, (lastProposedBlock, senders)) =>
+                        sendersSinceLastProposalIndex.add(
```

Comment:
How do we know whether another validator saw this block based on this logic? Unless we look into the justifications of other blocks, we can only learn what we saw since our own last proposed block.